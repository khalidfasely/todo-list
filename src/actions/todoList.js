import moment from 'moment';
import database from '../firebase/firebase';

// Add Todo Item
export const addTodoItem = (todoItem) => ({
    type: 'ADD_TODO_ITEM',
    todoItem
});

export const startAddTodoItem = (todoItemData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            content = '',
            active = false,
            time = moment(),
            timeOfDeactive = 0
        } = todoItemData;
        const todoItem = { content, active, time, timeOfDeactive };

        return database.ref(`users/${uid}/list`).push(todoItem).then((ref) => {
            dispatch(addTodoItem({
                id: ref.key,
                ...todoItem
            }));
        });
    };
};

// Deactive Item List
export const deactiveTodoItem = (id, updates) => ({
    type: 'DEACTIVE_TODO_ITEM',
    id,
    updates
});

export const startDeactiveTodoItem = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const updates = {
            active: false,
            timeOfDeactive: moment().format()
        };

        return database.ref(`users/${uid}/list/${id}`).update(updates).then(() => {
            dispatch(deactiveTodoItem(id, updates));
        });
    };
};

// Remove Item from List
export const removeTodoItem = (id) => ({
    type: 'REMOVE_TODO_ITEM',
    id
});

export const startRemoveTodoItem = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        
        return database.ref(`users/${uid}/list/${id}`).remove().then(() => {
            dispatch(removeTodoItem(id));
        });
    };
};

//Set Todo List
export const setTodoList = (todoList) => ({
    type: 'SET_TODO_LIST',
    todoList
});

export const startSetTodoList = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/list`)
            .once('value')
            .then((snapshot) => {
                const todoListData = [];

                snapshot.forEach((childSnapshot) => {
                    todoListData.push({
                      id: childSnapshot.key,
                      ...childSnapshot.val()
                    });
                });
                dispatch(setTodoList(todoListData));
            });
    };
};