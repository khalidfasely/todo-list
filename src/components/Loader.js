const Loader = () => (
    <div className="loader-page">
        <div className="loader-container">
            <div className="loader">
                <h2 data-text='Loading...'>Loading...</h2>
            </div>    
        </div>
    </div>
);

export default Loader;