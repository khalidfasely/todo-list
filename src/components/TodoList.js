import { connect } from 'react-redux';
import TodoItem from './TodoItem';

export const TodoList = ({ todoList = [] }) => {
    return (
        <div className='list-container'>
            {
                todoList.length !== 0 ?
                todoList.map(todoItem => <TodoItem  key={todoItem.id} {...todoItem} />) :
                <div className='list-empty'>Nothing Todo</div>
            }
        </div>
    );
};

const mapStateToProps = (state) => ({
    todoList: state.todoList
});

export default connect(mapStateToProps)(TodoList);