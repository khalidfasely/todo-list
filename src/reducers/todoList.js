// TodoList Reducer

const todoListReducerDefaultState = [];

export default (state = todoListReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_TODO_ITEM':
            return [
                ...state,
                action.todoItem
            ];
        case 'DEACTIVE_TODO_ITEM':
            return state.map(item => {
                if (item.id === action.id) {
                    return {
                        ...item,
                        ...action.updates
                    }
                } else {
                    return item;
                }
            });
        case 'REMOVE_TODO_ITEM':
            return state.filter(({ id }) => id !== action.id);
        case 'SET_TODO_LIST':
            return action.todoList;
        default:
            return state;
    }
};