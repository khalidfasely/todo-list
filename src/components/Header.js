import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => {
    return (
        <header className="header">
            <div className="header__logout-button-container">
                <button
                    className="header__logout-button"
                    data-testid='logout-button'
                    onClick={startLogout}
                >
                    Logout
                </button>
            </div>
            <h1 className="header__title">
                Todo List
            </h1>
        </header>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);