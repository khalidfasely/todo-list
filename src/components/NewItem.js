import { useState } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startAddTodoItem } from '../actions/todoList';

export const NewItem = ({ startAddTodoItem }) => {
    const [ newItem, setNewItem ] = useState('');
    const [ error, setError ] = useState('');
    //const [ todoList, setTodoList ] = useState([]);

    const onNewItemChange = (e) => {
        setNewItem(e.target.value);
    };

    const onAddItemToList = (e) => {
        setError('');
        e.preventDefault();
        const newTodoItem = {
            content: newItem,
            active: true,
            time: moment().format(),
            timeOfDeactive: undefined
        };
        if(newItem.replace(/\s/g, '')) {
            startAddTodoItem(newTodoItem);
            setNewItem('');
        } else {
            setError('Should fill the input!');
        };
    };

    //const onRemoveItem = (index) => {
    //    setTodoList([...todoList.filter((item, indexL) => index !== indexL)]);
    //};

    //const onDeactiveItem = (index) => {
    //    console.log(`Item number ${index} is done.`);
    //};

    return (
        <div className='form-container'>
            <form className='form' data-testid='form' onSubmit={onAddItemToList}>
                <div>
                    <input
                        data-testid='input'
                        placeholder="Add Something To do later!"
                        type='text'
                        maxLength={50}
                        value={newItem}
                        onChange={onNewItemChange}
                    />
                    <abbr title='Add Task ToDo'>
                        <button>+</button>
                    </abbr>
                </div>
                {error && <i className='form__error'>{error}</i>}
            </form>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startAddTodoItem: (newTodoItem) => dispatch(startAddTodoItem(newTodoItem))
});

export default connect(undefined, mapDispatchToProps)(NewItem);