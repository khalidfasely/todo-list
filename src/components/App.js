import Header from "./Header";
import NewItem from "./NewItem";
import '../firebase/firebase';
import { useEffect } from "react";
import database from "../firebase/firebase";

function App() {
  //useEffect(() => {
  //  database.ref('users')
  //  .once('value')
  //  .then((snapshot) => {
  //    const todoListData = [];
  //  
  //    snapshot.forEach((childSnapshot) => {
  //      todoListData.push({
  //        //id: childSnapshot.key,
  //        ...childSnapshot.val()
  //      });
  //    });
  //  
  //    console.log(todoListData);
  //  });
  //}, [])
  return (
    <div className="App">
      <Header />
      <NewItem />
    </div>
  );
};

export default App;
