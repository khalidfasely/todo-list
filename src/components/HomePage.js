import NewItem from "./NewItem";
import Header from "./Header";
import TodoList from "./TodoList";

const HomePage = () => {
    return (
        <div className="app">
            <div>
                <Header />
                <div className="sections-container">
                    <div className="content-container">
                        <NewItem />
                        <TodoList />
                    </div>
                </div>
            </div>
        </div>
    )
};

export default HomePage;