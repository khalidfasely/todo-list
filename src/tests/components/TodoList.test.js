import React from 'react';
import { render, screen } from '@testing-library/react';
import { TodoList } from '../../components/TodoList';
import todoList from '../fixtures/todoList';
jest.mock('../../components/TodoItem', () => () => <div>Todo Item Component</div>);

test('Should render TodoList component correctly with empty list', () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render TodoList component correctly', () => {
    const { asFragment } = render(<TodoList todoList={todoList} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render nothing todo paragraph if empty list', () => {
    render(<TodoList />);
    const nothingElement = screen.getByText(/nothing todo/i);
    expect(nothingElement).toBeInTheDocument();
});