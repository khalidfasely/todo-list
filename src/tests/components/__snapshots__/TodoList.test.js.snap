// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render TodoList component correctly 1`] = `
<DocumentFragment>
  <div
    class="list-container"
  >
    <div>
      Todo Item Component
    </div>
    <div>
      Todo Item Component
    </div>
    <div>
      Todo Item Component
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render TodoList component correctly with empty list 1`] = `
<DocumentFragment>
  <div
    class="list-container"
  >
    <div
      class="list-empty"
    >
      Nothing Todo
    </div>
  </div>
</DocumentFragment>
`;
